cli/cook/cli.py

```diff
@@ -0,0 +1,339 @@
+submit_parser.add_argument('--cpus', help='cpus to reserve for job', type=float)
+submit_parser.add_argument('--mem', help='memory to reserve for job', type=int)
+submit_parser.add_argument('--raw', '-r', help='raw job spec in json format', dest='raw', action='store_true')
+submit_parser.add_argument('command', nargs='?')
+submit_parser.add_argument('args', nargs=argparse.REMAINDER)
+
+
+def parse_raw_job_spec(job, r):
+    """
+    Parse a JSON string containing raw job data and merge with job template.
+    Job data can either be a dict of job attributes (indicating a single job),
+    or a list of dicts (indicating multiple jobs). In either case, the job attributes
+    are merged with (and override) the `job` template attributes.
+    Throws a ValueError if there is a problem parsing the data.
+    """
+    try:
+        content = json.loads(r)
+
+        if type(content) is dict:
+            return [merge_dicts(job, content)]
+        elif type(content) is list:
+            return [merge_dicts(job, c) for c in content]
+        else:
+            raise ValueError('invalid format for raw job')
+    except Exception:
+        raise ValueError('malformed JSON for raw job')
+
+
+def make_url(cluster, endpoint):
+    """Given a cluster and an endpoint, returns the corresponding full URL"""
+    return urljoin(cluster['url'], endpoint)
+
+
+def make_federated_request(clusters, make_request_fn, success_status, parse_response_fn,
+                           description, endpoint='rawscheduler'):
+    """
+    Attempts to make a request (via make_request_fn) to each cluster in clusters, until a cluster
+    returns a status code equal to success_status. If no cluster returns success status, throws.
+    """
+    for cluster in clusters:
+        try:
+            url = make_url(cluster, endpoint)
+            resp = make_request_fn(url)
+            logging.info('response from cook: %s' % resp.text)
+            if resp.status_code == success_status:
+                return parse_response_fn(resp, cluster)
+        except requests.exceptions.ConnectionError as ce:
+            logging.info(ce)
+    raise Exception('Unable to %s on any of the following clusters: %s' % (description, clusters))
+
+
+def safe_pop(d, key):
+    """If key is present in d, pops and returns the value. Otherwise, returns None."""
+    value = d.pop(key) if key in d else None
+    return value
+
+
+def submit(clusters, args):
+    """
+    Submits a job (or multiple jobs) to cook scheduler. Assembles a list of jobs, potentially getting data 
+    from configuration, the command line, and stdin.
+    """
+    logging.debug('submit args: %s' % args)
+
+    job = args
+    raw = safe_pop(job, 'raw')
+    command_from_command_line = safe_pop(job, 'command')
+    command_args = safe_pop(job, 'args')
+
+    if command_from_command_line:
+        commands = ['%s%s' % (command_from_command_line, (' ' + ' '.join(command_args)) if command_args else '')]
```

Comment:
This isn't safe.  You arguments might have spaces.  You should pass the array directly to `exec()` without shell interpretation.